README
JSON::RPC version 0.90
=================
NAME
    JSON::RPC - Perl implementation of JSON-RPC 1.1 protocol

DESCRIPTION
     JSON-RPC is a stateless and light-weight remote procedure call (RPC)
     protocol for inter-networking applications over HTTP. It uses JSON
     as the data format for of all facets of a remote procedure call,
     including all application data carried in parameters.

    quoted from <http://json-rpc.org/wd/JSON-RPC-1-1-WD-20060807.html>.

    This module was in JSON package on CPAN before. Now its interfaces was
    completely changed.

    The old modules - JSONRPC::Transport::HTTP and Apache::JSONRPC are
    deprecated. Please try to use JSON::RPC::Server and JSON::RPC::Client
    which support both JSON-RPC protocol version 1.1 and 1.0.

EXAMPLES
    CGI version.

     #--------------------------
     # In your application class
     package MyApp;
 
     use base qw(JSON::RPC::Procedure); # Perl 5.6 or more than
 
     sub echo : Public {    # new version style. called by clients
         # first argument is JSON::RPC::Server object.
         return $_[1];
     }
 
     sub sum : Public(a:num, b:num) { # sets value into object member a, b.
         my ($s, $obj) = @_;
         # return a scalar value or a hashref or an arryaref.
         return $obj->{a} + $obj->{b};
     }
 
     sub a_private_method : Private {
         # ... can't be called by client
     }
 
     sub sum_old_style {  # old version style. taken as Public
         my ($s, @arg) = @_;
        return $arg[0] + $arg[1];
     }
 
     #--------------------------
     # In your triger script.
     use JSON::RPC::Server::CGI;
     use MyApp;
 
     # simple
      JSON::RPC::Server::CGI->dispatch('MyApp')->handle();
 
     # or 
     JSON::RPC::Server::CGI->dispatch([qw/MyApp FooBar/])->handle();
 
     # or INFO_PATH version
     JSON::RPC::Server::CGI->dispatch({'/Test' => 'MyApp'})->handle();
 
     #--------------------------
     # Client
     use JSON::RPC::Client;
 
     my $client = new JSON::RPC::Client;

     my $uri = 'http://www.example.com/jsonrpc/Test';
     my $obj = {
        method  => 'sum', # or 'MyApp.sum'
        params  => [10, 20],
     };
 
     my $res = $client->call( $uri, $obj )
 
     if($res){
        if ($res->is_error) {
            print "Error : ", $res->error_message;
        }
        else {
            print $res->result;
        }
     }
     else {
        print $client->status_line;
     }
 
     # or
 
     $client->prepare($uri, ['sum', 'echo']);
     print $client->sum(10, 23);
 
    See to JSON::RPC::Server::CGI, JSON::RPC::Server::Daemon,
    JSON::RPC::Server::Apache JSON::RPC::Client and JSON::RPC::Procedure.

ABOUT NEW VERSION
    supports JSON-RPC protocol v1.1
TODO
    Document
    Examples
    More Tests
AUTHOR
    Makamaka Hannyaharamitu, <makamaka[at]cpan.org>

COPYRIGHT AND LICENSE
    Copyright 2007 by Makamaka Hannyaharamitu

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.


Changes
Changes
=======

1.05 2014-10-07T01:57:38Z
    - Work better with batch requests: Now responses are properly returned
      as list as long as the request is a list as well (Arata Makoto)

1.04 2014-05-29T11:42:19Z
    - Work with notifications, which are procedures without an id
      (Kirill Bogdanov)


1.03 2012 Jun 15
    - Allow die \%hash from handlers (ka2u)

1.02 2012 Jun 14
    - Fix the finalization of the response (0xAF)

1.01 2011 Dec 01
    - Use require instead of Class::Load. Subclass in your app if you need
      that kind of ability, but it seems unlikely that we'd need anything
      more than eval require here.

1.00 2011 Nov 22
    - No code change.

1.00_02  2011 Nov 16
    - JSON::RPC::Dispatcher already existed on CPAN. s/Dispatcher/Dispatch/g

1.00_01  2011 Nov 16
    - New maintainer, completely new code for PSGI apps and JSON RPC 2.0
    - If you are using old JSON::RPC code (up to 0.96), DO NOT EXPECT
      YOUR CODE TO WORK. THIS VERSION IS BACKWARDS INCOMPATIBLE

---------- Old Change Log -----------

Revision history for Perl extension JSON::RPC.

###### In the next large version up ######################################

  * JSON::RPC::Server::Apache will be renamed to JSON::RPC::Server::Apache2
      and split into another distribution.
  * JSON::RPC::Server::Apache really supports apache 1.3x
      and the maintainer will be changed.

##########################################################################

0.96  Mon Feb 25 11:06:25 2008
    - JSON::RPC::Server::FastCGI was split into the independent distribution.
        the new maintainer is Faiz Kazi.
    - JSON::RPC::Server::Apache was renamed to JSON::RPC::Server::Apache2
        it will split into another distribution.
    - added and updated docs.


0.95  Fri Feb 15 16:01:04 2008
    - sample codes were indexed...!
        fixed package for avoiding the indexer.


0.94  Fri Feb 15 15:16:32 2008
    - no change but examples was forgotten.


0.93  Fri Feb 15 14:46:17 2008
    - added example codes.
    - now AUTOLOAD method after prepare() can support built-in methods.
        $client -> __VERSION__ ( => $client calls the name 'VERSION' procedure )
    - Your application can set subroutines allowable by allowable_procedure().
        (Thanks to seagull's suggestion)
    - JSON::RPC::Server::Apache config supports 'return_die_message'.
    - require LWP::UserAgent 2.001 or later.


0.92  Thu Feb 14 13:12:40 2008
    - modified the JSON::RPC::Client prepare mode to check response errors.
        (Thanks to Colin Meyer)
    - fixed retrieve_json_from_get in JSON::RPC::Server::CGI.
    - implemented JSON::RPC::Server::Apache to support the GET method call.
    - fixed JSON::RPC::ReturnObject decoding JSON data with utf8 mode.
        ($client -> call() ->result will return Unicode characters (if contained).)
    - added some descriptions to the JSON::RPC::Client pod.


0.91  Wed Dec 19 15:51:53 2007
    - JSON::RPC::Client used JSON::PP.
    - added create_json_coder() to JSON::RPC::Client.


0.90  Wed Dec 19 13:26:15 2007
    - Now default JSON coder is JSON! (1.99 or later)
    - added JSON::RPC::Server::FastCGI written by Faiz Kazi, thanks!
    - added JSONRPC for guide to this distribution.
    - cleaned up JSON::RPC::Server::CGI
    - added create_json_coder() to JSON::RPC::Server.
    - modified J::R::Server::* as base.pm does not work well in Perl 5.005


0.01  Mon May 21 14:18:33 2007
    - original version

lib/JSON/RPC.pm
package JSON::RPC;

use strict;

$JSON::RPC::VERSION = '0.95';


1;
__END__

=pod

=head1 NAME

JSON::RPC - Perl implementation of JSON-RPC 1.1 protocol

=head1 DESCRIPTION

 JSON-RPC is a stateless and light-weight remote procedure call (RPC)
 protocol for inter-networking applications over HTTP. It uses JSON
 as the data format for of all facets of a remote procedure call,
 including all application data carried in parameters.

quoted from L<http://json-rpc.org/wd/JSON-RPC-1-1-WD-20060807.html>.


This module was in JSON package on CPAN before.
Now its interfaces was completely changed.

The old modules - L<JSONRPC::Transport::HTTP> and L<Apache::JSONRPC> are deprecated.
Please try to use JSON::RPC::Server and JSON::RPC::Client which support both JSON-RPC
protocol version 1.1 and 1.0.


=head1 EXAMPLES

CGI version.

 #--------------------------
 # In your application class
 package MyApp;
 
 use base qw(JSON::RPC::Procedure); # Perl 5.6 or more than
 
 sub echo : Public {    # new version style. called by clients
     # first argument is JSON::RPC::Server object.
     return $_[1];
 }
 
 
 sub sum : Public(a:num, b:num) { # sets value into object member a, b.
     my ($s, $obj) = @_;
     # return a scalar value or a hashref or an arryaref.
     return $obj->{a} + $obj->{b};
 }
 
  
 sub a_private_method : Private {
     # ... can't be called by client
 }
 
 
 sub sum_old_style {  # old version style. taken as Public
     my ($s, @arg) = @_;
    return $arg[0] + $arg[1];
 }
 
 
 #--------------------------
 # In your triger script.
 use JSON::RPC::Server::CGI;
 use MyApp;
 
 # simple
  JSON::RPC::Server::CGI->dispatch('MyApp')->handle();
 
 # or 
 JSON::RPC::Server::CGI->dispatch([qw/MyApp FooBar/])->handle();
 
 # or INFO_PATH version
 JSON::RPC::Server::CGI->dispatch({'/Test' => 'MyApp'})->handle();
 
 #--------------------------
 # Client
 use JSON::RPC::Client;
 
 my $client = new JSON::RPC::Client;

 my $uri = 'http://www.example.com/jsonrpc/Test';
 my $obj = {
    method  => 'sum', # or 'MyApp.sum'
    params  => [10, 20],
 };
 
 my $res = $client->call( $uri, $obj )
 
 if($res){
    if ($res->is_error) {
        print "Error : ", $res->error_message;
    }
    else {
        print $res->result;
    }
 }
 else {
    print $client->status_line;
 }
 
 # or
 
 $client->prepare($uri, ['sum', 'echo']);
 print $client->sum(10, 23);
 

See to L<JSON::RPC::Server::CGI>, L<JSON::RPC::Server::Daemon>, L<JSON::RPC::Server::Apache>
L<JSON::RPC::Client> and L<JSON::RPC::Procedure>.


=head1 ABOUT NEW VERSION

=over

=item supports JSON-RPC protocol v1.1


=back

=head1 TODO

=over

=item Document

=item Examples

=item More Tests


=back


=head1 AUTHOR

Makamaka Hannyaharamitu, E<lt>makamaka[at]cpan.orgE<gt>


=head1 COPYRIGHT AND LICENSE

Copyright 2007-2008 by Makamaka Hannyaharamitu

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 

=cut


